datastore/modelassertmanager.go

```diff
@@ -0,0 +1,154 @@
+FROM modelassertion
+WHERE model_id=$1
+`
+
+// ModelAssertion holds the model assertion details in the local database
+type ModelAssertion struct {
+	ID           int       `json:"id"`
+	ModelID      int       `json:"model_id"`
+	KeypairID    int       `json:"keypair_id"`
+	Series       int       `json:"series"`
+	Architecture string    `json:"architecture"`
+	Revision     int       `json:"revision"`
+	Gadget       string    `json:"gadget"`
+	Kernel       string    `json:"kernel"`
+	Store        string    `json:"store"`
+	Created      time.Time `json:"created"`
+	Modified     time.Time `json:"modified"`
+}
+
+// CreateModelAssertTable creates the database table for a model assertion
+func (db *DB) CreateModelAssertTable() error {
+	_, err := db.Exec(createModelAssertTableSQL)
+	return err
+}
+
+// CreateModelAssert adds a model assertion record to allow generation of a signed assertion
+func (db *DB) CreateModelAssert(m ModelAssertion) (int, error) {
+	var createdID int
+	err := db.QueryRow(createModelAssertSQL, m.ModelID, m.KeypairID, m.Series, m.Architecture, m.Revision, m.Gadget, m.Kernel, m.Store).Scan(&createdID)
+	if err != nil {
+		log.Printf("Error creating the model assertion: %v\n", err)
+	}
+	return createdID, err
+}
+
+// UpdateModelAssert updates the model assertion details
+func (db *DB) UpdateModelAssert(m ModelAssertion) error {
+
+	var err error
+
+	_, err = db.Exec(updateModelAssertSQL, m.ID, m.ModelID, m.KeypairID, m.Series, m.Architecture, m.Revision, m.Gadget, m.Kernel, m.Store, time.Now().UTC())
+
+	if err != nil {
+		log.Printf("Error updating the model assertion: %v\n", err)
+	}
+
+	return err
+
+}
+
+// UpsertModelAssert creates or updates the model assertion headers
+func (db *DB) UpsertModelAssert(m ModelAssertion) error {
+
+	var err error
+
+	if ok := validateModelAssertion(m); ok != nil {
+		return ok
+	}
+
+	if m.ID > 0 {
+		err = db.UpdateModelAssert(m)
+	} else {
+		_, err = db.CreateModelAssert(m)
+	}
+
+	return err
+
+}
+
+// GetModelAssert fetches the model assertion
+func (db *DB) GetModelAssert(modelID int) (ModelAssertion, error) {
+	m := ModelAssertion{}
+	err := db.QueryRow(getModelAssertSQL, modelID).Scan(&m.ID, &m.ModelID, &m.KeypairID, &m.Series, &m.Architecture, &m.Revision, &m.Gadget, &m.Kernel, &m.Store, &m.Created, &m.Modified)
+	if err != nil {
+		return m, err
+	}
+
+	return m, nil
+}
+
+func validateModelAssertion(m ModelAssertion) error {
+	if m.ModelID <= 0 {
+		return errors.New("Model must be provided")
+	}
+	if m.KeypairID <= 0 {
+		return errors.New("Signing Key must be provided")
+	}
+	if m.Series < 16 {
+		return errors.New("Series must be at least 16")
+	}
+
+	if err := validateNotEmpty("Architecture", m.Architecture); err != nil {
+		return err
+	}
+	if err := validateNotEmpty("Gadget", m.Gadget); err != nil {
+		return err
+	}
+	if err := validateNotEmpty("Kernel", m.Kernel); err != nil {
+		return err
+	}
+	err := validateNotEmpty("Store", m.Store)
```

Comment:
more cosmetic: I would use same formula as before if err := .. etc and having a return nil at the end, or if you prefer, the opposite:
return validateNotEmpty(...

Current seems a confusing thing of both.

Said this, no matter if left so :)

Reply:
That's how I had it originally, but got a lint error on travis, so had to change it!